scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,184 @@
+
+_GLOBAL_CONFIG = {
+    "SCIPY_ARRAY_API": SCIPY_ARRAY_API,
+    "SCIPY_DEVICE": SCIPY_DEVICE,
+}
+
+
+def compliance_scipy(*arrays):
+    """Raise exceptions on known-bad subclasses.
+
+    The following subclasses are not supported and raise and error:
+    - `np.ma.MaskedArray`
+    - `numpy.matrix`
+    - Any array-like which is not Array API compatible
+    """
+    for array in arrays:
+        if isinstance(array, np.ma.MaskedArray):
+            raise TypeError("'numpy.ma.MaskedArray' are not supported")
+        elif isinstance(array, np.matrix):
+            raise TypeError("'numpy.matrix' are not supported")
+        elif not array_api_compat.is_array_api_obj(array):
+            raise TypeError("Only support Array API compatible arrays")
+        elif array.dtype is np.dtype('O'):
+            raise TypeError('object arrays are not supported')
+
+
+def _check_finite(array, xp):
+    """Check for NaNs or Infs."""
+    msg = "array must not contain infs or NaNs"
+    try:
+        if not xp.all(xp.isfinite(array)):
+            raise ValueError(msg)
+    except TypeError:
+        raise ValueError(msg)
+
+
+def array_namespace(*arrays):
+    """Get the array API compatible namespace for the arrays xs.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays used to infer the common namespace.
+
+    Returns
+    -------
+    namespace : module
+        Common namespace.
+
+    Notes
+    -----
+    Thin wrapper around `array_api_compat.array_namespace`.
+
+    1. Check for the global switch: SCIPY_ARRAY_API. This can also be accessed
+       dynamically through ``_GLOBAL_CONFIG['SCIPY_ARRAY_API']``.
+    2. `compliance_scipy` raise exceptions on known-bad subclasses. See
+       it's definition for more details.
+
+    When the global switch is False, it defaults to the `numpy` namespace.
+    In that case, there is no compliance check. This is a convenience to
+    ease the adoption. Otherwise, arrays must comply with the new rules.
+    """
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        # here we could wrap the namespace if needed
+        return array_api_compat_numpy
+
+    arrays = [array for array in arrays if array is not None]
+
+    compliance_scipy(*arrays)
+
+    return array_api_compat.array_namespace(*arrays)
+
+
+def as_xparray(
+    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=True
+):
+    """Drop-in replacement for `np.asarray`.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases.
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
```

Comment:
AFAIU, see the comment bellow, the idea is to use normal NumPy and then convert back to `numpy.array_api`.

(This is from sklearn.)

Reply:
AFAIU, see the comment bellow, the idea is to use normal NumPy and then convert back to `numpy.array_api`.

(This is from sklearn.)

Reply:
Ah, I see what is going on there. That seems like the wrong thing to do to me. The whole point of using `numpy.array_api` is to test with a strictly-compliant implementation of the standard, so avoiding that by converting to regular `numpy` kinda defeats the purpose. I think we want `numpy.array_api.Array` instances to take the non-numpy-specific route here.

There's a second issue in this line I just noticed: the `array_api_compat` name is incorrect because the module is now vendored, it should be `"scipy._lib.array_api_compat.array_api_compat.numpy"`.

Reply:
It's aliased `import scipy._lib.array_api_compat.array_api_compat as array_api_compat`

Reply:
Yes, but you're comparing exact module names; `__name__` doesn't care about what the import alias is. You can `print(xp)` to see it.

I think this is the change to apply:

```diff
diff --git a/scipy/_lib/_array_api.py b/scipy/_lib/_array_api.py
index d7d184731..fbc39bd21 100644
--- a/scipy/_lib/_array_api.py
+++ b/scipy/_lib/_array_api.py
@@ -109,7 +109,7 @@ def as_xparray(
     """
     if xp is None:
         xp = array_namespace(array)
-    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
+    if xp.__name__ in {"numpy", "scipy._lib.array_api_compat.array_api_compat.numpy"}:
         # Use NumPy API to support order
         if copy is True:
             array = np.array(array, order=order, dtype=dtype)
```